### ISSUE TITLE
Special keys like Ctrl, Alt, Meta, Esc, F1

### ISSUE THREAD

**Issue by eckende** [florisboard/florisboard#229]
> First of all, I love what you're doing here - a functional keyboard that also looks nice and professional. Thank you for your work!

Currently, there is no way to access special keys like Ctrl, Alt, Meta, (Esc, the function keys, Home, End, Page up/down etc.).
There are many ways this or a subset of this could be implemented:
1. create another keyboard page (like ?123 for numbers).
2. add a keyboard subtype/keyboard layout to be used in situations where these keys are useful
3. option to show Ctrl, Alt, maybe Esc, Meta in the normal layout
4. add button to smartbar that opens keyboard page
5. this could work nicely with #196 keyboard customization

For reference have a look at the Hacker's Keyboard in landscape mode and press the Fn key.
([GitHub Repo](https://github.com/klausw/hackerskeyboard), [Play Store](https://play.google.com/store/apps/details?id=org.pocketworkstation.pckeyboard), [F-Droid](https://f-droid.org/en/packages/org.pocketworkstation.pckeyboard/)

**Comment by patrickgold** [comment]
Thanks for your suggestion!

This gave me an idea how I can redesign the selection keyboard (#216). If I make the selection keyboard quickly accessible, I could design a selection keyboard which has all special keys you mentioned plus the arrow keys in one thing, somewhat going in the direction the BeHe Keyboard does (https://play-lh.googleusercontent.com/ezirmAGOFY6mD8FdWomTb_qM6pJS-KW8MOs6fQvzQFZauZ529uG1fk2DnM1ZsG7zOQY=w720-h310). This would give the selection keyboard a purpose and reason to be used.

I will add it on my todo list, probably for v0.5.0 for now.

**Comment by rancidfrog** [comment]
Also,
Having the 'Enter' key always available in special keys area would be useful, as sometimes enter is replaced by send even though it is also possible to use enter key, but it is just not available

**Comment by patrickgold** [comment]
Yes, I agree that the enter key should also be present on the selection keyboard. Will keep this in mind when redesigning the selection keyboard!

**Comment by rancidfrog** [comment]
Delete key, is also useful for nee selection keyboard.

Looking forward to future implementation.
Once implemented will not be needing hackers keyboard anymore
Keep the great work going.

**Comment by yashpalgoyal1304** [comment]
> Delete key, is also useful for new selection keyboard.

hi, as u seem to be using delete key often, what do you think about having a (forward) delete key as suggested here: #356

**Comment by mlvnt** [comment]
This is a little less know keyboard [Technical Keyboard](https://play.google.com/store/apps/details?id=nextapp.inputmethod.latin&hl=en&gl=US) I'm using for years, and the only one that nailed special keys support. There is a lot of inspiration that can be borrowed from it:

![](https://user-images.githubusercontent.com/18243897/109363458-4cddda00-7895-11eb-873e-680bbba58cd8.jpg)
![](https://user-images.githubusercontent.com/18243897/109363462-4e0f0700-7895-11eb-9202-e5b1006666a9.jpg)
![](https://user-images.githubusercontent.com/18243897/109363463-4e0f0700-7895-11eb-9767-f35f059b018f.jpg)

One another feature, kind of related is this one:

![](https://user-images.githubusercontent.com/18243897/109363594-9f1efb00-7895-11eb-8bd7-d53a85025413.jpg)

Super useful, have not seen any other mobile keyboard even including a way to insert tabs

**Comment by justinesmithies** [comment]
> > First of all, I love what you're doing here - a functional keyboard that also looks nice and professional. Thank you for your work!
> 
> Currently, there is no way to access special keys like Ctrl, Alt, Meta, (Esc, the function keys, Home, End, Page up/down etc.).
> There are many ways this or a subset of this could be implemented:
> 
>     1. create another keyboard page (like ?123 for numbers).
> 
>     2. add a keyboard subtype/keyboard layout to be used in situations where these keys are useful
> 
>     3. option to show Ctrl, Alt, maybe Esc, Meta in the normal layout
> 
>     4. add button to smartbar that opens keyboard page
> 
>     5. this could work nicely with #196 keyboard customization
> 
> 
> For reference have a look at the Hacker's Keyboard in landscape mode and press the Fn key.
> ([GitHub Repo](https://github.com/klausw/hackerskeyboard), [Play Store](https://play.google.com/store/apps/details?id=org.pocketworkstation.pckeyboard), [F-Droid](https://f-droid.org/en/packages/org.pocketworkstation.pckeyboard/))
> 
> Another keyboard that implements this is the BeHe keyboard, press the three dots.
> ([GitHub Repo](https://github.com/VladThodo/behe-keyboard), [Play Store](https://play.google.com/store/apps/details?id=com.vlath.keyboard), [F-Droid](https://f-droid.org/en/packages/com.vlath.keyboard/))

I've just switched from excuse my language Gboard to FlorisBoard and already i'd give it 100% but adding the above features will make this the best keyboard by far ! Thanks for the amazing work !

**Comment by another-sapiens** [comment]
This is awesome! And especially useful for apps like KDE Connect.

**Comment by atikattar1104** [comment]
I Think There Must Be An Option To Implement 'Full 5-Row Layout' Of Hacker's Keyboard. Currently, This Is My Theme For  Florisboard :- 
![Screenshot_2022-09-17-10-25-01-651_com android chrome](https://user-images.githubusercontent.com/93785261/190841097-b75b116b-b3a2-49e3-ba83-1ff24a08c642.jpg)
And This Is The Full 5-Row Layout I Use Of Hacker's Keyboard :- 
![Screenshot_2022-09-17-10-27-21-866_com android chrome](https://user-images.githubusercontent.com/93785261/190841167-fe8c7e01-ea54-4d1d-af58-5256b07324e4.jpg)
And This Is How It Looks Normally :- 
![Screenshot_2022-09-17-10-28-19-526_com android chrome](https://user-images.githubusercontent.com/93785261/190841201-6c452f51-6578-4df7-864f-69cb309f1e5e.jpg)
Now I Would Definitely Love To Have The Full 5-Row Layout Which Have My Theme. The Layout Of ABC, Number, Special Keys Should Be Implemented In Florisboard, But, The Smartbar (Which Suggests Word And The Extendable One) Should Be There (The One Which alpha04 Have Right Now). Even When We Select 'Full 5-Row Layout' From Florisboard Settings App.

Thanks!

**Comment by atikattar1104** [comment]
And Perhaps, You Can Also Use Same Key Commands Which Hacker's Keyboard Use, Which Are Being Sent To OS, In New Implemented Version Of Florisboard. So That, Florisboard Will Be Just As Compatible As Hacker's Keyboard When It Comes To Using It On Special Softwares (E.g. :- Termux, ConnectBot, Limbo PC Emulator, DosBox, Etc.).

**Comment by noahgitsham** [comment]
Not sure on the status of this, but on keyboards such as hacker's keyboard, it always felt cramped with all those modifiers. It would be nice to have the option to bind them to swipe/hold gestures as well

**Comment by lepras** [comment]
following, this has the most duplicates lol, prob a sign.

this specific one iz most active so should close others

**Comment by user-1-2-3-4-5** [comment]
>For reference have a look at the Hacker's Keyboard in landscape mode and press the Fn key.
([GitHub Repo](https://github.com/klausw/hackerskeyboard), [Play Store](https://play.google.com/store/apps/details?id=org.pocketworkstation.pckeyboard), [F-Droid](https://f-droid.org/en/packages/org.pocketworkstation.pckeyboard/))
>
>Another keyboard that implements this is the BeHe keyboard, press the three dots.
([GitHub Repo](https://github.com/VladThodo/behe-keyboard), [Play Store](https://play.google.com/store/apps/details?id=com.vlath.keyboard), [F-Droid](https://f-droid.org/en/packages/com.vlath.keyboard/))

[Unexpected Keyboard](https://github.com/Julow/Unexpected-Keyboard) has also been released since, supports all of the keys mentioned in the title.

**Comment by AdityaMH** [comment]
Need to be pinned

**Comment by lm41** [comment]
> Need to be pinned

No there are other priorities right now.

**Comment by atikattar1104** [comment]
> No there are other priorities right now.

Agreed. Base Level Stability Should Be 1st Priority. Additional Features + Their Stability Should Be Later.
